### PR TITLE
refactored extracting self contained schemas

### PR DESCRIPTION
**Pull request by AhmetTanakol**
- Refactored extracting self contained schemas
- Added unit tests to get containment properties

### CODE REVIEW

File: packages/editor/test/services/containment.property.test.ts [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+    expect(properties[1].label).toEqual('b');
+    const aCopy = JSON.parse(JSON.stringify(schema.definitions.a));
+    aCopy.id = '#' + (schema.properties.elements.items as JsonSchema).anyOf[0].$ref;
+    const bCopy = JSON.parse(JSON.stringify(schema.definitions.b));
+    bCopy.id = '#' + (schema.properties.elements.items as JsonSchema).anyOf[1].$ref;
+    expect(properties[0].schema).toMatchObject(aCopy);
+    expect(properties[1].schema).toMatchObject(bCopy);
+    done();
+  });
+  test('support for array references', done => {
```
Comment (edgarmueller) [inline packages/editor/test/services/containment.property.test.ts]: Move all tests which make use of `links` to different test file.

File: tslint.json [hunk 2]
```diff
@@ -63,7 +63,7 @@
         "no-unsafe-any": false,
         "no-unsafe-finally": true,
         "no-unused-expression": true,
-        "no-unused-variable": true,
+        "no-unused-variable": [true, {"ignore-pattern": "^_"}],
```
Comment (edgarmueller) [inline tslint.json]: Remove pattern, keep warning

File: packages/editor/src/services/schema.service.impl.ts [hunk 3]
```diff
@@ -336,6 +336,87 @@ const getSchemaIdForObject = (object: Object, modelMapping: ModelMapping): strin
...
+  },              allRefs);
+};
+
+/**
+ * Calculate references that are used in parentSchema and add copy them into schema
+ * @param {JsonSchema} parentSchema
+ * @param {JsonSchema} schema
+ * @returns {JsonSchema}
+ */
+export const calculateSchemaReferences = (parentSchema: JsonSchema,
```
Comment (edgarmueller) [inline packages/editor/src/services/schema.service.impl.ts]: Describe parameters and return type

File: packages/editor/src/services/schema.service.impl.ts [hunk 4]
```diff
@@ -336,6 +336,87 @@ const getSchemaIdForObject = (object: Object, modelMapping: ModelMapping): strin
...
+    }
+    if (!extractedReferences[schemaRefValue.uri]) {
+      extractedReferences[schemaRefValue.uri] = schemaRefValue.uri;
+      prev = _.omitBy(prev, value => value.uri === schemaRefValue.uri);
+    }
+    if (refs !== undefined) {
+      findReferencePaths(parentSchema, prev, refs, extractedReferences);
+    }
+    return prev;
+  },              allRefs);
```
Comment (edgarmueller) [inline packages/editor/src/services/schema.service.impl.ts]: Formatting seems to be off a bit here

File: packages/editor/src/services/schema.service.impl.ts [hunk 5]
```diff
@@ -336,6 +336,87 @@ const getSchemaIdForObject = (object: Object, modelMapping: ModelMapping): strin
...
+
+/**
+ * Finding required definitions from parentSchema by using schema refs
+ *
+ * @param parentSchema The root schema
+ * @param allRefs All the refs of root schema
+ * @param schemaRefs The current schema refs
+ * @param extractedReferences Contains the definition attributes for the current schema
+ */
+const findReferencePaths = (parentSchema: JsonSchema,
```
Comment (edgarmueller) [inline packages/editor/src/services/schema.service.impl.ts]: Describe return type

File: packages/editor/example/config.ts [hunk 6]
```diff
@@ -25,4 +25,4 @@ export const detailSchemata = {
   '#task': taskView,
   '#user': userView,
   '#userGroup': userGroupView,
-};
+};
```
Comment (edgarmueller) [inline packages/editor/example/config.ts]: Add newline

File: packages/editor/example/schema.ts [hunk 7]
```diff
@@ -81,4 +81,4 @@ export const taskSchema = {
       'additionalProperties': false
     }
   }
-};
+};
```
Comment (edgarmueller) [inline packages/editor/example/schema.ts]: Add newline

File: packages/editor/test/services/service.test.ts [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    };
+    const schema = {
+      type: 'object',
+      properties: {
+        bar: {type: 'string'}
+      }
+    };
+    const calculatedSchemaReferences = calculateSchemaReferences(parentSchema, schema);
+    expect(calculatedSchemaReferences).toMatchObject(schema);
+    done();
```
[the remark this answers is not in the document]
Comment (AhmetTanakol) [reply packages/editor/test/services/service.test.ts]: I checked the documentation again, we don't need it

**Comment by edgarmueller** [comment]
@AhmetTanakol The build is probably failing because of node type definitions, see https://github.com/eclipsesource/jsonforms/pull/947/commits/4a3ff69446ed31e7a95cf7f8c724a09aeeb90696

**Comment by AhmetTanakol** [comment]
I kept those files for my own testing purpose. Currently, the main example is the one that Lucas added ( User and Task Editor )

**Comment by edgarmueller** [comment]
I'd propose to merge this for now. We can still updated the examples later, but we should get this PR in as follow-up work is based on this.

**Comment by eneufeld** [comment]
@edgarmueller fine with me, as long as we keep track of removing unneeded example files

**Comment by edgarmueller** [comment]
See https://github.com/eclipsesource/jsonforms/issues/956

**Pull request merged**